File: curriculum/02_key_value_store/go-impl/internal/kvstore/store.go
package kvstore

import (
	"encoding/json"
	"errors"
	"sort"
	"time"
	"unicode/utf8"

	"sync"
)

const (
	ErrInvalidKey          = "INVALID_KEY"
	ErrKeyTooLong          = "KEY_TOO_LONG"
	ErrInvalidTTL          = "INVALID_TTL"
	ErrKeyNotFound         = "KEY_NOT_FOUND"
	ErrValueTooLarge       = "VALUE_TOO_LARGE"
	ErrStoreFull           = "STORE_FULL"
	ErrMemoryLimitExceeded = "MEMORY_LIMIT_EXCEEDED"
	ErrInvalidJSON         = "INVALID_JSON"
	ErrInvalidLimit        = "INVALID_LIMIT"
)

const (
	defaultMaxKeyBytes    = 512
	defaultMaxValueBytes  = 1 << 20
	defaultMaxKeys        = 100_000
	defaultMaxMemoryBytes = 256 << 20
	minTTLSeconds         = 1
	maxTTLSeconds         = 30 * 24 * 60 * 60
	entryOverheadBytes    = 64
)

type Config struct {
	MaxKeyBytes    int
	MaxValueBytes  int
	MaxKeys        int
	MaxMemoryBytes int
}

type Store struct {
	mu                 sync.RWMutex
	entries            map[string]StoredEntry
	config             Config
	now                func() time.Time
	approxMemoryBytes  int
	commandsProcessed  uint64
	expiredKeysRemoved uint64
}

type StoredEntry struct {
	Key         string
	Value       any
	CreatedAt   time.Time
	UpdatedAt   time.Time
	ExpiresAt   *time.Time
	ApproxBytes int
}

type EntryView struct {
	Key        string     `json:"key"`
	Value      any        `json:"value"`
	TTLSeconds *int64     `json:"ttlSeconds"`
	ExpiresAt  *time.Time `json:"expiresAt"`
}

type Pair struct {
	Key   string `json:"key"`
	Value any    `json:"value"`
}

type MGetItem struct {
	Key   string `json:"key"`
	Value any    `json:"value"`
	Found bool   `json:"found"`
}

type Health struct {
	Status             string `json:"status"`
	KeyCount           int    `json:"keyCount"`
	ApproxMemoryBytes  int    `json:"approxMemoryBytes"`
	CommandsProcessed  uint64 `json:"commandsProcessed"`
	ExpiredKeysRemoved uint64 `json:"expiredKeysRemoved"`
}

type DomainError struct {
	Code    string
	Message string
}

func (e DomainError) Error() string { return e.Message }

func IsCode(err error, code string) bool {
	var domainErr DomainError
	return errors.As(err, &domainErr) && domainErr.Code == code
}

func NewStore(config Config, now func() time.Time) *Store {
	if now == nil {
		now = time.Now
	}
	if config.MaxKeyBytes == 0 {
		config.MaxKeyBytes = defaultMaxKeyBytes
	}
	if config.MaxValueBytes == 0 {
		config.MaxValueBytes = defaultMaxValueBytes
	}
	if config.MaxKeys == 0 {
		config.MaxKeys = defaultMaxKeys
	}
	if config.MaxMemoryBytes == 0 {
		config.MaxMemoryBytes = defaultMaxMemoryBytes
	}
	return &Store{entries: map[string]StoredEntry{}, config: config, now: now}
}

func (s *Store) Set(key string, value any, ttlSeconds int64) (*time.Time, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.commandsProcessed++
	s.removeExpiredLocked(s.now())
	expiresAt, approx, err := s.validateWriteLocked(key, value, ttlSeconds, map[string]struct{}{key: {}})
	if err != nil {
		return nil, err
	}
	now := s.now()
	old := s.entries[key]
	s.approxMemoryBytes += approx - old.ApproxBytes
	created := old.CreatedAt
	if created.IsZero() {
		created = now
	}
	s.entries[key] = StoredEntry{Key: key, Value: cloneJSON(value), CreatedAt: created, UpdatedAt: now, ExpiresAt: expiresAt, ApproxBytes: approx}
	return expiresAt, nil
}

func (s *Store) Get(key string) (EntryView, bool) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.commandsProcessed++
	now := s.now()
	if !s.validStoredKeyLocked(key, now) {
		return EntryView{}, false
	}
	return s.viewLocked(s.entries[key], now), true
}

func (s *Store) Delete(keys []string) int {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.commandsProcessed++
	now := s.now()
	deleted := 0
	for _, key := range keys {
		if s.validStoredKeyLocked(key, now) {
			s.removeLocked(key)
			deleted++
		}
	}
	return deleted
}

func (s *Store) Expire(key string, ttlSeconds int64) (bool, *time.Time, error) {
	if err := validateTTL(ttlSeconds); err != nil {
		return false, nil, err
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	s.commandsProcessed++
	now := s.now()
	if !s.validStoredKeyLocked(key, now) {
		return false, nil, DomainError{Code: ErrKeyNotFound, Message: "key not found"}
	}
	expiresAt := now.Add(time.Duration(ttlSeconds) * time.Second)
	entry := s.entries[key]
	entry.ExpiresAt = &expiresAt
	entry.UpdatedAt = now
	s.entries[key] = entry
	return true, &expiresAt, nil
}

func (s *Store) TTL(key string) int64 {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.commandsProcessed++
	now := s.now()
	if !s.validStoredKeyLocked(key, now) {
		return -2
	}
	entry := s.entries[key]
	if entry.ExpiresAt == nil {
		return -1
	}
	return remainingSeconds(*entry.ExpiresAt, now)
}

func (s *Store) Persist(key string) (bool, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.commandsProcessed++
	now := s.now()
	if !s.validStoredKeyLocked(key, now) {
		return false, DomainError{Code: ErrKeyNotFound, Message: "key not found"}
	}
	entry := s.entries[key]
	if entry.ExpiresAt == nil {
		return false, nil
	}
	entry.ExpiresAt = nil
	entry.UpdatedAt = now
	s.entries[key] = entry
	return true, nil
}

func (s *Store) Keys(prefix string, limit int) []string {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.commandsProcessed++
	s.removeExpiredLocked(s.now())
	keys := make([]string, 0, len(s.entries))
	for key := range s.entries {
		if prefix == "" || len(key) >= len(prefix) && key[:len(prefix)] == prefix {
			keys = append(keys, key)
		}
	}
	sort.Strings(keys)
	if limit > 0 && len(keys) > limit {
		return keys[:limit]
	}
	return keys
}

func (s *Store) MGet(keys []string) []MGetItem {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.commandsProcessed++
	now := s.now()
	items := make([]MGetItem, 0, len(keys))
	for _, key := range keys {
		if !s.validStoredKeyLocked(key, now) {
			items = append(items, MGetItem{Key: key, Value: nil, Found: false})
			continue
		}
		items = append(items, MGetItem{Key: key, Value: cloneJSON(s.entries[key].Value), Found: true})
	}
	return items
}

func (s *Store) MSet(items []Pair, ttlSeconds int64) (*time.Time, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.commandsProcessed++
	now := s.now()
	s.removeExpiredLocked(now)
	seen := make(map[string]struct{}, len(items))
	var expiresAt *time.Time
	newMemory := s.approxMemoryBytes
	newKeys := len(s.entries)
	approxs := make([]int, len(items))
	for i, item := range items {
		if _, exists := seen[item.Key]; exists {
			return nil, DomainError{Code: ErrInvalidKey, Message: "duplicate key in mset"}
		}
		seen[item.Key] = struct{}{}
		expiry, approx, err := s.validateWriteLocked(item.Key, item.Value, ttlSeconds, seen)
		if err != nil {
			return nil, err
		}
		if i == 0 {
			expiresAt = expiry
		}
		old, exists := s.entries[item.Key]
		if !exists {
			newKeys++
		}
		newMemory += approx - old.ApproxBytes
		approxs[i] = approx
	}
	if newKeys > s.config.MaxKeys {
		return nil, DomainError{Code: ErrStoreFull, Message: "store key limit exceeded"}
	}
	if newMemory > s.config.MaxMemoryBytes {
		return nil, DomainError{Code: ErrMemoryLimitExceeded, Message: "memory limit exceeded"}
	}
	for i, item := range items {
		old := s.entries[item.Key]
		created := old.CreatedAt
		if created.IsZero() {
			created = now
		}
		var itemExpiresAt *time.Time
		if ttlSeconds > 0 {
			expiry := now.Add(time.Duration(ttlSeconds) * time.Second)
			itemExpiresAt = &expiry
		}
		s.entries[item.Key] = StoredEntry{Key: item.Key, Value: cloneJSON(item.Value), CreatedAt: created, UpdatedAt: now, ExpiresAt: itemExpiresAt, ApproxBytes: approxs[i]}
	}
	s.approxMemoryBytes = newMemory
	return expiresAt, nil
}

func (s *Store) FlushDB() int {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.commandsProcessed++
	deleted := len(s.entries)
	s.entries = map[string]StoredEntry{}
	s.approxMemoryBytes = 0
	return deleted
}

func (s *Store) Health() Health {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.removeExpiredLocked(s.now())
	return Health{Status: "ok", KeyCount: len(s.entries), ApproxMemoryBytes: s.approxMemoryBytes, CommandsProcessed: s.commandsProcessed, ExpiredKeysRemoved: s.expiredKeysRemoved}
}

func (s *Store) validateWriteLocked(key string, value any, ttlSeconds int64, batchKeys map[string]struct{}) (*time.Time, int, error) {
	if err := s.validateKey(key); err != nil {
		return nil, 0, err
	}
	if ttlSeconds != 0 {
		if err := validateTTL(ttlSeconds); err != nil {
			return nil, 0, err
		}
	}
	valueBytes, err := json.Marshal(value)
	if err != nil {
		return nil, 0, DomainError{Code: ErrInvalidJSON, Message: "value is not JSON serializable"}
	}
	if len(valueBytes) > s.config.MaxValueBytes {
		return nil, 0, DomainError{Code: ErrValueTooLarge, Message: "value is too large"}
	}
	if _, exists := s.entries[key]; !exists && len(s.entries) >= s.config.MaxKeys && len(batchKeys) == 1 {
		return nil, 0, DomainError{Code: ErrStoreFull, Message: "store key limit exceeded"}
	}
	var expiresAt *time.Time
	if ttlSeconds > 0 {
		expiry := s.now().Add(time.Duration(ttlSeconds) * time.Second)
		expiresAt = &expiry
	}
	approx := len(key) + len(valueBytes) + entryOverheadBytes
	old := s.entries[key]
	if s.approxMemoryBytes+approx-old.ApproxBytes > s.config.MaxMemoryBytes {
		return nil, 0, DomainError{Code: ErrMemoryLimitExceeded, Message: "memory limit exceeded"}
	}
	return expiresAt, approx, nil
}

func (s *Store) validateKey(key string) error {
	if key == "" || !utf8.ValidString(key) {
		return DomainError{Code: ErrInvalidKey, Message: "key must be a non-empty UTF-8 string"}
	}
	if len(key) > s.config.MaxKeyBytes {
		return DomainError{Code: ErrKeyTooLong, Message: "key is too long"}
	}
	return nil
}

func validateTTL(ttlSeconds int64) error {
	if ttlSeconds < minTTLSeconds || ttlSeconds > maxTTLSeconds {
		return DomainError{Code: ErrInvalidTTL, Message: "ttlSeconds must be between 1 and 2592000"}
	}
	return nil
}

func (s *Store) validStoredKeyLocked(key string, now time.Time) bool {
	entry, ok := s.entries[key]
	if !ok {
		return false
	}
	if isExpired(entry, now) {
		s.removeLocked(key)
		s.expiredKeysRemoved++
		return false
	}
	return true
}

func (s *Store) removeExpiredLocked(now time.Time) {
	for key, entry := range s.entries {
		if isExpired(entry, now) {
			s.removeLocked(key)
			s.expiredKeysRemoved++
		}
	}
}

func (s *Store) removeLocked(key string) {
	entry := s.entries[key]
	delete(s.entries, key)
	s.approxMemoryBytes -= entry.ApproxBytes
	if s.approxMemoryBytes < 0 {
		s.approxMemoryBytes = 0
	}
}

func (s *Store) viewLocked(entry StoredEntry, now time.Time) EntryView {
	var ttl *int64
	if entry.ExpiresAt != nil {
		remaining := remainingSeconds(*entry.ExpiresAt, now)
		ttl = &remaining
	}
	return EntryView{Key: entry.Key, Value: cloneJSON(entry.Value), TTLSeconds: ttl, ExpiresAt: entry.ExpiresAt}
}

func isExpired(entry StoredEntry, now time.Time) bool {
	return entry.ExpiresAt != nil && !entry.ExpiresAt.After(now)
}

func remainingSeconds(expiresAt time.Time, now time.Time) int64 {
	remaining := int64(expiresAt.Sub(now).Seconds())
	if remaining < 0 {
		return 0
	}
	return remaining
}

func cloneJSON(value any) any {
	bytes, err := json.Marshal(value)
	if err != nil {
		return value
	}
	var cloned any
	if err := json.Unmarshal(bytes, &cloned); err != nil {
		return value
	}
	return cloned
}

File: curriculum/02_key_value_store/node-impl/src/store.ts
import type { EntryView, JsonValue, Pair } from './types';

const DEFAULT_MAX_KEY_BYTES = 512;
const DEFAULT_MAX_VALUE_BYTES = 1 << 20;
const DEFAULT_MAX_KEYS = 100_000;
const DEFAULT_MAX_MEMORY_BYTES = 256 << 20;
const ENTRY_OVERHEAD_BYTES = 64;
const MAX_TTL_SECONDS = 30 * 24 * 60 * 60;
const NANOS_PER_SECOND = 1_000_000_000n;
// MINOR-003 (docs/code_review.md): /health used to run a full O(n) removeExpired()
// sweep on every single call. Monitoring systems poll /health frequently (often
// every few seconds), so this made the cheapest, most-polled endpoint in the
// service the one deliberately made O(n) in the resident key count. Rate-limiting
// the sweep keeps keyCount/expiredKeysRemoved accurate to within this window
// (lazy per-key cleanup on get/delete/etc. already guarantees expired keys are
// never returned to clients in between sweeps — see RF-011) while making repeat
// /health polls within the window O(1).
const HEALTH_SWEEP_MIN_INTERVAL_NANOS = 1_000_000_000n; // 1s

export enum ErrorCode {
  InvalidCommand = 'INVALID_COMMAND',
  InvalidJson = 'INVALID_JSON',
  InvalidKey = 'INVALID_KEY',
  KeyTooLong = 'KEY_TOO_LONG',
  InvalidTtl = 'INVALID_TTL',
  InvalidLimit = 'INVALID_LIMIT',
  KeyNotFound = 'KEY_NOT_FOUND',
  ValueTooLarge = 'VALUE_TOO_LARGE',
  StoreFull = 'STORE_FULL',
  MemoryLimitExceeded = 'MEMORY_LIMIT_EXCEEDED'
}

export class DomainError extends Error {
  constructor(public readonly code: ErrorCode, message: string) {
    super(message);
    this.name = 'DomainError';
  }
}

export interface StoreConfig {
  maxKeyBytes?: number;
  maxValueBytes?: number;
  maxKeys?: number;
  maxMemoryBytes?: number;
}

interface RequiredStoreConfig {
  maxKeyBytes: number;
  maxValueBytes: number;
  maxKeys: number;
  maxMemoryBytes: number;
}

interface StoredEntry {
  value: JsonValue;
  createdAtNanos: bigint;
  updatedAtNanos: bigint;
  expiresAtNanos: bigint | null;
  expiresAtDate: Date | null;
  approxBytes: number;
}

interface WritePlan {
  expiresAtNanos: bigint | null;
  expiresAtDate: Date | null;
  approxBytes: number;
}

export class KeyValueStore {
  private readonly entries = new Map<string, StoredEntry>();
  private readonly config: RequiredStoreConfig;
  private approxMemoryBytes = 0;
  private commandsProcessed = 0;
  private expiredKeysRemoved = 0;
  private lastHealthSweepNanos: bigint | null = null;

  constructor(config: StoreConfig = {}, private readonly nowNanos: () => bigint = process.hrtime.bigint) {
    this.config = {
      maxKeyBytes: config.maxKeyBytes ?? DEFAULT_MAX_KEY_BYTES,
      maxValueBytes: config.maxValueBytes ?? DEFAULT_MAX_VALUE_BYTES,
      maxKeys: config.maxKeys ?? DEFAULT_MAX_KEYS,
      maxMemoryBytes: config.maxMemoryBytes ?? DEFAULT_MAX_MEMORY_BYTES
    };
  }

  set(key: string, value: JsonValue, ttlSeconds?: number): { key: string; stored: true; expiresAt?: string } {
    this.commandsProcessed += 1;
    this.removeExpired(this.nowNanos());
    const plan = this.validateWrite(key, value, ttlSeconds);
    const now = this.nowNanos();
    const existing = this.entries.get(key);
    this.approxMemoryBytes += plan.approxBytes - (existing?.approxBytes ?? 0);
    this.entries.set(key, {
      value: cloneJson(value),
      createdAtNanos: existing?.createdAtNanos ?? now,
      updatedAtNanos: now,
      expiresAtNanos: plan.expiresAtNanos,
      expiresAtDate: plan.expiresAtDate,
      approxBytes: plan.approxBytes
    });
    return plan.expiresAtDate === null
      ? { key, stored: true }
      : { key, stored: true, expiresAt: plan.expiresAtDate.toISOString() };
  }

  get(key: string): EntryView | null {
    this.commandsProcessed += 1;
    const now = this.nowNanos();
    if (!this.validStoredKey(key, now)) {
      return null;
    }
    const entry = this.entries.get(key);
    return entry === undefined ? null : this.view(key, entry, now);
  }

  delete(keys: readonly string[]): { deleted: number } {
    this.commandsProcessed += 1;
    const now = this.nowNanos();
    let deleted = 0;
    for (const key of keys) {
      if (this.validStoredKey(key, now)) {
        this.removeKey(key);
        deleted += 1;
      }
    }
    return { deleted };
  }

  expire(key: string, ttlSeconds: number): { updated: boolean; ttlSeconds: number; expiresAt: string } {
    this.validateKey(key);
    validateTtl(ttlSeconds);
    this.commandsProcessed += 1;
    const now = this.nowNanos();
    if (!this.validStoredKey(key, now)) {
      throw new DomainError(ErrorCode.KeyNotFound, 'key not found');
    }
    const entry = this.entries.get(key);
    if (entry === undefined) {
      throw new DomainError(ErrorCode.KeyNotFound, 'key not found');
    }
    const expiresAtNanos = now + BigInt(ttlSeconds) * NANOS_PER_SECOND;
    const expiresAtDate = new Date(Date.now() + ttlSeconds * 1000);
    entry.expiresAtNanos = expiresAtNanos;
    entry.expiresAtDate = expiresAtDate;
    entry.updatedAtNanos = now;
    return { updated: true, ttlSeconds, expiresAt: expiresAtDate.toISOString() };
  }

  ttl(key: string): { ttlSeconds: number } {
    this.commandsProcessed += 1;
    const now = this.nowNanos();
    if (!this.validStoredKey(key, now)) {
      return { ttlSeconds: -2 };
    }
    const entry = this.entries.get(key);
    if (entry?.expiresAtNanos === null || entry === undefined) {
      return { ttlSeconds: -1 };
    }
    return { ttlSeconds: remainingSeconds(entry.expiresAtNanos, now) };
  }

  persist(key: string): { updated: boolean } {
    this.commandsProcessed += 1;
    const now = this.nowNanos();
    if (!this.validStoredKey(key, now)) {
      throw new DomainError(ErrorCode.KeyNotFound, 'key not found');
    }
    const entry = this.entries.get(key);
    if (entry === undefined) {
      throw new DomainError(ErrorCode.KeyNotFound, 'key not found');
    }
    const updated = entry.expiresAtNanos !== null;
    entry.expiresAtNanos = null;
    entry.expiresAtDate = null;
    entry.updatedAtNanos = now;
    return { updated };
  }

  keys(prefix = '', limit = 1000): { keys: string[]; count: number } {
    this.commandsProcessed += 1;
    this.removeExpired(this.nowNanos());
    const keys = [...this.entries.keys()].filter((key) => key.startsWith(prefix)).sort().slice(0, limit);
    return { keys, count: keys.length };
  }

  mget(keys: readonly string[]): { items: Array<{ key: string; value: JsonValue | null; found: boolean }> } {
    this.commandsProcessed += 1;
    const now = this.nowNanos();
    const items = keys.map((key) => {
      if (!this.validStoredKey(key, now)) {
        return { key, value: null, found: false };
      }
      const entry = this.entries.get(key);
      return { key, value: entry === undefined ? null : cloneJson(entry.value), found: entry !== undefined };
    });
    return { items };
  }

  mset(items: readonly Pair[], ttlSeconds?: number): { stored: number; expiresAt?: string } {
    this.commandsProcessed += 1;
    this.removeExpired(this.nowNanos());
    const seen = new Set<string>();
    const plans: WritePlan[] = [];
    let newKeyCount = this.entries.size;
    let newMemory = this.approxMemoryBytes;
    for (const item of items) {
      if (seen.has(item.key)) {
        throw new DomainError(ErrorCode.InvalidKey, 'duplicate key in mset');
      }
      seen.add(item.key);
      const plan = this.validateWrite(item.key, item.value, ttlSeconds);
      if (!this.entries.has(item.key)) {
        newKeyCount += 1;
      }
      newMemory += plan.approxBytes - (this.entries.get(item.key)?.approxBytes ?? 0);
      plans.push(plan);
    }
    if (newKeyCount > this.config.maxKeys) {
      throw new DomainError(ErrorCode.StoreFull, 'store key limit exceeded');
    }
    if (newMemory > this.config.maxMemoryBytes) {
      throw new DomainError(ErrorCode.MemoryLimitExceeded, 'memory limit exceeded');
    }
    const now = this.nowNanos();
    items.forEach((item, index) => {
      const existing = this.entries.get(item.key);
      const plan = plans[index];
      if (plan === undefined) {
        return;
      }
      this.entries.set(item.key, {
        value: cloneJson(item.value),
        createdAtNanos: existing?.createdAtNanos ?? now,
        updatedAtNanos: now,
        expiresAtNanos: plan.expiresAtNanos,
        expiresAtDate: plan.expiresAtDate,
        approxBytes: plan.approxBytes
      });
    });
    this.approxMemoryBytes = newMemory;
    const expiresAt = plans.find((plan) => plan.expiresAtDate !== null)?.expiresAtDate?.toISOString();
    return expiresAt === undefined ? { stored: items.length } : { stored: items.length, expiresAt };
  }

  flushdb(): { deleted: number } {
    this.commandsProcessed += 1;
    const deleted = this.entries.size;
    this.entries.clear();
    this.approxMemoryBytes = 0;
    return { deleted };
  }

  health(): { status: 'ok'; keyCount: number; approxMemoryBytes: number; commandsProcessed: number; expiredKeysRemoved: number } {
    const now = this.nowNanos();
    if (this.lastHealthSweepNanos === null || now - this.lastHealthSweepNanos >= HEALTH_SWEEP_MIN_INTERVAL_NANOS) {
      this.removeExpired(now);
      this.lastHealthSweepNanos = now;
    }
    return {
      status: 'ok',
      keyCount: this.entries.size,
      approxMemoryBytes: this.approxMemoryBytes,
      commandsProcessed: this.commandsProcessed,
      expiredKeysRemoved: this.expiredKeysRemoved
    };
  }

  validateKey(key: string): void {
    if (key.length === 0) {
      throw new DomainError(ErrorCode.InvalidKey, 'key must be non-empty');
    }
    if (Buffer.byteLength(key, 'utf8') > this.config.maxKeyBytes) {
      throw new DomainError(ErrorCode.KeyTooLong, 'key is too long');
    }
  }

  private validateWrite(key: string, value: JsonValue, ttlSeconds?: number): WritePlan {
    this.validateKey(key);
    if (ttlSeconds !== undefined) {
      validateTtl(ttlSeconds);
    }
    const serialized = JSON.stringify(value);
    const serializedBytes = Buffer.byteLength(serialized, 'utf8');
    if (serializedBytes > this.config.maxValueBytes) {
      throw new DomainError(ErrorCode.ValueTooLarge, 'value is too large');
    }
    if (!this.entries.has(key) && this.entries.size >= this.config.maxKeys) {
      throw new DomainError(ErrorCode.StoreFull, 'store key limit exceeded');
    }
    const approxBytes = Buffer.byteLength(key, 'utf8') + serializedBytes + ENTRY_OVERHEAD_BYTES;
    const oldBytes = this.entries.get(key)?.approxBytes ?? 0;
    if (this.approxMemoryBytes + approxBytes - oldBytes > this.config.maxMemoryBytes) {
      throw new DomainError(ErrorCode.MemoryLimitExceeded, 'memory limit exceeded');
    }
    if (ttlSeconds === undefined) {
      return { approxBytes, expiresAtNanos: null, expiresAtDate: null };
    }
    return {
      approxBytes,
      expiresAtNanos: this.nowNanos() + BigInt(ttlSeconds) * NANOS_PER_SECOND,
      expiresAtDate: new Date(Date.now() + ttlSeconds * 1000)
    };
  }

  private validStoredKey(key: string, now: bigint): boolean {
    const entry = this.entries.get(key);
    if (entry === undefined) {
      return false;
    }
    if (entry.expiresAtNanos !== null && entry.expiresAtNanos <= now) {
      this.removeKey(key);
      this.expiredKeysRemoved += 1;
      return false;
    }
    return true;
  }

  private removeExpired(now: bigint): void {
    for (const [key, entry] of this.entries) {
      if (entry.expiresAtNanos !== null && entry.expiresAtNanos <= now) {
        this.removeKey(key);
        this.expiredKeysRemoved += 1;
      }
    }
  }

  private removeKey(key: string): void {
    const entry = this.entries.get(key);
    if (entry !== undefined) {
      this.approxMemoryBytes = Math.max(0, this.approxMemoryBytes - entry.approxBytes);
      this.entries.delete(key);
    }
  }

  private view(key: string, entry: StoredEntry, now: bigint): EntryView {
    return {
      key,
      value: cloneJson(entry.value),
      ttlSeconds: entry.expiresAtNanos === null ? null : remainingSeconds(entry.expiresAtNanos, now),
      expiresAt: entry.expiresAtDate?.toISOString() ?? null
    };
  }
}

function validateTtl(ttlSeconds: number): void {
  if (!Number.isInteger(ttlSeconds) || ttlSeconds <= 0 || ttlSeconds > MAX_TTL_SECONDS) {
    throw new DomainError(ErrorCode.InvalidTtl, 'ttlSeconds must be between 1 and 2592000');
  }
}

function remainingSeconds(expiresAt: bigint, now: bigint): number {
  const remaining = (expiresAt - now) / NANOS_PER_SECOND;
  return Number(remaining < 0n ? 0n : remaining);
}

function cloneJson(value: JsonValue): JsonValue {
  return JSON.parse(JSON.stringify(value)) as JsonValue;
}
